In [1]:
#Hide code cells when saving as HTML
from IPython.display import display
from IPython.display import HTML
import IPython.core.display as di # Example: di.display_html('<h3>%s:</h3>' % str, raw=True)

# This line will hide code by default when the notebook is exported as HTML
di.display_html('<script>jQuery(function() {if (jQuery("body.notebook_app").length == 0) { jQuery(".input_area").toggle(); jQuery(".prompt").toggle();}});</script>', raw=True)

# This line will add a button to toggle visibility of code blocks, for use with the HTML export version
di.display_html('''<button onclick="jQuery('.input_area').toggle(); jQuery('.prompt').toggle();">Toggle code</button>''', raw=True)

Toggle code

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib.dates as dates
import plotly.graph_objects as go
from pandas_profiling import ProfileReport

In [3]:
#import functions created in .py
import load_data 
import clean_data
import transform_data
import eda
import geda

In [140]:
import warnings

def fxn():
    warnings.warn("deprecated", DeprecationWarning)

with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    fxn()

# Homicidios INEGI

De las 64 variables que tiene la base original, seleccionamos solo las de interés. 

Abajo mostramos la información general de la base.

In [5]:
##################################################  CARGAR BASE  #############################################################
#Update data file name
datafile_name = 'inegi_homicidios_1990-2018.csv'

#Update path to data file
path_to_datafile = '/home/bj/Documents/Xaber_ODP/DB/Homicidios/INEGI_Defunciones'


#we load the database
inegi_homicidios = load_data.load_df(datafile_name, path_to_datafile, ',')

/home/bj/.local/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3254: DtypeWarning:

Columns (33,43) have mixed types.Specify dtype option on import or set low_memory=False.



In [6]:
#Mostramos algunas lineas del archivo
#inegi_homicidios.columns
#inegi_homicidios.head()

In [7]:
########################################  LIMPIAR NOMBRE VARIABLES  ######################################################
# Estandarizacion de los nombres de las columnas
clean_homicidios = clean_data.clean_variables_names(inegi_homicidios)
#clean_homicidios.columns
#clean_homicidios.head()
homicidios_vars = clean_homicidios[['ent_regis', 'mun_regis', 'dia_regis', 'mes_regis', 'anio_regis',
                                   'ent_ocurr', 'mun_ocurr', 'loc_ocurr', 'dia_ocurr', 'mes_ocurr', 'anio_ocur', 
                                   'ent_ocules', 'mun_ocules', 'loc_ocules']]

In [8]:
############################################## INFORMACION GENERAL ##################################################
# Informacion general de la base
eda.df_variables_info(homicidios_vars)
# Clasificación de las variables por tipo
[numeric_variables, categorical_variables, dates_variables, string_variables] = eda.info_type_of_vars(homicidios_vars)
# Imprime a pantalla las variables por tipo
eda.print_info_type_of_vars(numeric_variables, categorical_variables, dates_variables, string_variables)

Number of variables: 14
Number of observations: 497754
Type of variables in the dataframe: 
 int64      8
float64    6
dtype: int64 

Numericas:
 ['ent_regis' 'mun_regis' 'dia_regis' 'mes_regis' 'anio_regis' 'ent_ocurr'
 'mun_ocurr' 'loc_ocurr' 'dia_ocurr' 'mes_ocurr' 'anio_ocur' 'ent_ocules'
 'mun_ocules' 'loc_ocules'] 
 Total:  14 

Categoricas:
 [] 
 Total:  0 

Formato Fecha: 
 [] 
 Total:  0 

Foramto texto: 
 [] 
 Total:  0 



In [9]:
#profile = ProfileReport(homicidios_vars, title='Pandas Profiling Report Homicidios', html={'style':{'full_width':True}})

In [10]:
#eda.descriptive_stats_for_categorical_vars(homicidios_vars, string_variables)

In [11]:
eda.descriptive_stats_for_numeric_vars(homicidios_vars, numeric_variables)

ent_regis      mun_regis      dia_regis      mes_regis  \
n_observations    497754.000000  497754.000000  328776.000000  497754.000000   
max                   32.000000     570.000000      99.000000      12.000000   
min                    1.000000       1.000000       1.000000       1.000000   
mean                  15.449371      50.252275      23.431695       6.599786   
std_dev                7.432461      70.347191      25.355667       3.434177   
25%                    9.000000      12.000000       9.000000       4.000000   
median                15.000000      32.000000      17.000000       7.000000   
75%                   20.000000      58.000000      26.000000      10.000000   
kurtosis              -0.561198      20.241798       4.286251      -1.194664   
skewness               0.361286       3.964865       2.293053      -0.027509   
n_unique_values       32.000000     486.000000      32.000000      12.000000   
%_missings             0.000000       0.000000      33.948095       0.000000   
Top1_most_common      15.000000      39.000000      99.000000      12.000000   
Top2_most_common       8.000000      15.000000      18.000000       5.000000   
Top3_most_common      12.000000       1.000000      17.000000      10.000000   

                     anio_regis      ent_ocurr      mun_ocurr      loc_ocurr  \
n_observations    497754.000000  497754.000000  497754.000000  134105.000000   
max                 2018.000000      99.000000     999.000000    9999.000000   
min                   90.000000       1.000000       1.000000       1.000000   
mean                1541.025701      15.894934      58.981041     995.598076   
std_dev              824.245737       9.506173     115.024784    2789.381071   
25%                 1998.000000       9.000000      10.000000       1.000000   
median              2009.000000      15.000000      32.000000       1.000000   
75%                 2014.000000      20.000000      59.000000       8.000000   
kurtosis              -0.591348      27.298315      40.682813       4.768253   
skewness              -1.186774       3.459103       5.820329       2.566230   
n_unique_values       29.000000      36.000000     538.000000     704.000000   
%_missings             0.000000       0.000000       0.000000      73.057976   
Top1_most_common    2018.000000      15.000000      37.000000       1.000000   
Top2_most_common    2017.000000       8.000000       4.000000    7777.000000   
Top3_most_common    2011.000000      12.000000       1.000000    9999.000000   

                      dia_ocurr      mes_ocurr      anio_ocur     ent_ocules  \
n_observations    375891.000000  497754.000000  497754.000000  183135.000000   
max                   99.000000      99.000000    9999.000000      99.000000   
min                    1.000000       1.000000       3.000000       1.000000   
mean                  16.429694       7.136533    1553.774798      81.825779   
std_dev               11.704557       7.955730     889.801594      22.359077   
25%                    8.000000       4.000000    1998.000000      88.000000   
median                16.000000       7.000000    2009.000000      88.000000   
75%                   24.000000      10.000000    2014.000000      88.000000   
kurtosis              19.169757     104.997140      11.688303       5.071584   
skewness               2.955544       9.305907       0.376191      -2.583322   
n_unique_values       32.000000      13.000000     143.000000      34.000000   
%_missings            24.482576       0.000000       0.000000      63.207729   
Top1_most_common       1.000000      12.000000    2018.000000      88.000000   
Top2_most_common      25.000000       5.000000    2017.000000      99.000000   
Top3_most_common      20.000000       8.000000    2011.000000      11.000000   

                     mun_ocules     loc_ocules  
n_observations    183135.000000  160072.000000  
max                  999.000000    9999.000000  
min                    1.000000 

Notemos que:

   - La variable entidad de registro (*ent_regis*) tiene 32 valires únicos correspondientes a los 32 estados, mientras que la variable entidad de ocurrencia (*ent_ocurr*) tiene 36 valores únicos.
   - La variables día de registro (*dia_regis*) y dia de ocurrencia (*dia_ocurr*) tienen 32 valores únicos: uno por cada día del mes más otro adicional para los valores desconocidos. Por su parte, la variable dia de ocurrencia tiene 24% de valores faltantes.
   - La variable mes de registro (*mes_regis*) tiene 12 valores únicos correspondientes a cada mes, mientras que la variable de mes de ocurrencia (*mes_ocurr*) tiene un valor adicional para los valores de mes desconocido.
   - La variable año de registro (*anio_regis*) tiene exactamente 29 valores únicos, uno por cada año desde 1990 al 2018, aunque tiene 34% de valores faltantes. La variable (*anio_ocur*) tiene 143 valores únicos y aunque no tiene valores faltantes, algunos valores podrían ser erróneos.
   - La variable localidad de ocurrencia (*loc_ocurr*: Localidad donde ocurrió el hecho vital) tiene 73% de valores faltantes.
   - Las variables de entidad, municipio y localidad donde ocurrio la lessión (*ent_ocules*, *mun_ocules*, y *loc_ocules*, respectivamente) tienen 63%, 63% y 68% de valores faltantes.
   
A continuación tenemos más detalle sobre los valores únicos de cada variable.

In [12]:
#-------------- Ano de registro vs Ano de ocurrencia
ano_regis = homicidios_vars['anio_regis'].unique()
ano_ocurr = homicidios_vars['anio_ocur'].unique()
print("Ano de registro: \n", np.sort(ano_regis), "\n", "Ano de ocurrencia: \n", np.sort(ano_ocurr))
# Anio_ocurr tiene un rango {1914...2018, 9999}

Ano de registro: 
 [  90   91   92   93   94   95   96   97 1998 1999 2000 2001 2002 2003
 2004 2005 2006 2007 2008 2009 2010 2011 2012 2013 2014 2015 2016 2017
 2018] 
 Ano de ocurrencia: 
 [   3    4    5    7    8    9   10   11   14   19   20   24   28   29
   40   43   44   45   46   47   50   51   52   54   55   57   58   59
   63   65   68   69   71   72   73   74   75   76   77   78   79   80
   81   82   83   84   85   86   87   88   89   90   91   92   93   94
   95   96   97   99 1900 1901 1912 1914 1916 1920 1925 1927 1929 1930
 1936 1937 1938 1943 1946 1948 1949 1952 1953 1954 1956 1957 1958 1959
 1960 1961 1962 1963 1964 1965 1966 1967 1968 1969 1970 1972 1973 1974
 1975 1976 1977 1978 1979 1980 1981 1982 1983 1984 1985 1986 1987 1988
 1989 1990 1991 1992 1993 1994 1995 1996 1997 1998 1999 2000 2001 2002
 2003 2004 2005 2006 2007 2008 2009 2010 2011 2012 2013 2014 2015 2016
 2017 2018 9999]


In [25]:
#-------------- Mes de registro vs Mes de ocurrencia
mes_regis = homicidios_vars['mes_regis'].unique()
mes_ocurr = homicidios_vars['mes_ocurr'].unique()
print("Mes de registro: \n", np.sort(mes_regis), "\n", "Mes de ocurrencia: \n", np.sort(mes_ocurr))

Mes de registro: 
 [ 1  2  3  4  5  6  7  8  9 10 11 12] 
 Mes de ocurrencia: 
 [ 1  2  3  4  5  6  7  8  9 10 11 12 99]


In [26]:
#-------------- Dia de registro vs Dia de ocurrencia
dia_regis = homicidios_vars['dia_regis'].unique()
dia_ocurr = homicidios_vars['dia_ocurr'].unique()
print("Dia de registro: \n", np.sort(dia_regis), "\n", "Dia de ocurrencia: \n", np.sort(dia_ocurr))

Dia de registro: 
 [ 1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17. 18.
 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 99. nan] 
 Dia de ocurrencia: 
 [ 1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17. 18.
 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 99. nan]


In [15]:
#-------------- Entidad de registro vs Entidad de ocurrencia
entidad_regis = homicidios_vars['ent_regis'].unique()
entidad_ocurr = homicidios_vars['ent_ocurr'].unique()
print("Entidades de registro: \n", np.sort(entidad_regis), "\n", "Entidades de ocurrencia: \n", np.sort(entidad_ocurr))

Entidades de registro: 
 [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32] 
 Entidades de ocurrencia: 
 [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32 33 34 35 99]


## Limpieza de variables

#### Año de ocurrencia

Notamos que la variable año de registro (*anio_regis*) tiene abreviados a 2 cifras los años 1990 a 1997.

Para la variable año de ocurrencia, vamos a considerar del mismo año las variables 90 y 1990, 91 y 1991 y así sucesivamente hasta el año 97 y 1997. Los demás valores no se considerarán.

In [107]:
by_anio_regis = homicidios_vars.anio_regis.value_counts().sort_index()
#print(by_anio_regis)
by_anio_ocurr = homicidios_vars.anio_ocur.value_counts().sort_index()
#print(by_anio_ocurr)

homicidios_vars.loc[homicidios_vars.anio_ocur == 90, 'anio_ocur'] = 1990
homicidios_vars.loc[homicidios_vars.anio_ocur == 91, 'anio_ocur'] = 1991
homicidios_vars.loc[homicidios_vars.anio_ocur == 92, 'anio_ocur'] = 1992
homicidios_vars.loc[homicidios_vars.anio_ocur == 93, 'anio_ocur'] = 1993
homicidios_vars.loc[homicidios_vars.anio_ocur == 94, 'anio_ocur'] = 1994
homicidios_vars.loc[homicidios_vars.anio_ocur == 95, 'anio_ocur'] = 1995
homicidios_vars.loc[homicidios_vars.anio_ocur == 96, 'anio_ocur'] = 1996
homicidios_vars.loc[homicidios_vars.anio_ocur == 97, 'anio_ocur'] = 1997
#homicidios_vars.anio_ocur.loc[(homicidios_vars['anio_ocur'] == 90)] = 1990
#homicidios_vars['anio_ocur'] = np.where(homicidios_vars['anio_ocur'] == 90, 1990, homicidios_vars['anio_ocur'])

by_anio_ocurr = homicidios_vars.anio_ocur.value_counts().sort_index()
#print(by_anio_ocurr[105:135,])

1990    11568
1991    12741
1992    13886
1993    13634
1994    15771
1995    15516
1996    14435
1997    13463
1998    13617
1999    12184
2000    10726
2001    10304
2002    10095
2003    10034
2004     9368
2005    10005
2006    10437
2007     8870
2008    14272
2009    19814
2010    26124
2011    27463
2012    25950
2013    22575
2014    19672
2015    20369
2016    24662
2017    32080
2018    35522
9999      789
Name: anio_ocur, dtype: int64


/home/bj/anaconda3/envs/Xaber/lib/python3.8/site-packages/pandas/core/indexing.py:965: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



#### Entidades de ocurrencia

Se muestra el porcentaje de valores faltantes para la variable entidad de ocurrencia por año de ocurrencia

In [104]:
edo_ocur_missing = homicidios_vars.loc[homicidios_vars.ent_ocurr > 32].groupby(['anio_ocur']).size()
edo_ocur_missing = pd.DataFrame(edo_ocur_missing)
edo_ocur_missing.columns = ['#_missings']
edo_ocur_missing['%_missings'] = edo_ocur_missing['#_missings']/(by_anio_ocurr)
edo_ocur_missing.loc[1990:2018]

#_missings  %_missings
anio_ocur                        
1990               66    0.005705
1991               81    0.006357
1992              115    0.008282
1993              110    0.008068
1994              111    0.007038
1995              118    0.007605
1996               82    0.005681
1997               90    0.006685
1998               65    0.004773
1999               77    0.006320
2000               73    0.006806
2001               70    0.006793
2002               59    0.005844
2003               28    0.002791
2004                3    0.000320
2005                9    0.000900
2006                5    0.000479
2007                4    0.000451
2008                2    0.000140
2010                1    0.000038
2011                4    0.000146
2012               12    0.000462
2013                1    0.000044
2014                9    0.000458
2015                7    0.000344
2016               30    0.001216
2017              400    0.012469
2018             1934    0.054445

#### Localidad de ocurrencia

Se muestra el porcentaje de valores faltantes para la variable localidad de ocurrencia por año de ocurrencia.

In [138]:
loc_ocur_missing = homicidios_vars.loc[(homicidios_vars.loc_ocurr.isna()) | (homicidios_vars.loc_ocurr == 99)].groupby(['anio_ocur']).size()
loc_ocur_missing = pd.DataFrame(loc_ocur_missing)
loc_ocur_missing.columns = ['#_missings']
loc_ocur_missing['%_missings'] = loc_ocur_missing['#_missings']/(by_anio_ocurr)
loc_ocur_missing.loc[1990:2018]

#_missings  %_missings
anio_ocur                        
1990            11567    0.999914
1991            12740    0.999922
1992            13886    1.000000
1993            13632    0.999853
1994            15769    0.999873
1995            15516    1.000000
1996            14434    0.999931
1997            13463    1.000000
1998            13613    0.999706
1999            12182    0.999836
2000            10722    0.999627
2001            10299    0.999515
2002            10090    0.999505
2003            10032    0.999801
2004             9366    0.999787
2005            10003    0.999800
2006            10430    0.999329
2007             8861    0.998985
2008            14255    0.998809
2009            19784    0.998486
2010            26035    0.996593
2011            27354    0.996031
2012            25749    0.992254
2013            21749    0.963411
2014                5    0.000254
2015                7    0.000344
2016                6    0.000243
2017                7    0.000218
2018                7    0.000197

In [139]:
#homicidios_vars.loc[(homicidios_vars.anio_ocur == 2014)]

In [ ]:
#'ent_regis' 'mun_regis' 'ent_resid' 'mun_resid' 'tloc_resid' 'ent_ocurr'
# 'mun_ocurr' 'lista_bas' 'sexo' 'edad' 'mes_ocurr' 'anio_ocur' 'mes_regis'
# 'anio_regis' 'mes_nacim' 'anio_nacim' 'ocupacion' 'escolarida'
# 'edo_civil' 'presunto' 'ocurr_trab' 'lugar_ocur' 'necropsia' 'asist_medi'
# 'sitio_ocur' 'cond_cert' 'cert_nomed' 'nacionalid' 'derechohab'
# 'edad_agru' 'maternas' 'dis_re_oax' 'dia_ocurr' 'dia_nacim' 'embarazo'
# 'rel_emba' 'horas' 'minutos' 'capitulo' 'grupo' 'lista1' 'vio_fami'
# 'tloc_ocurr' 'dia_regis' 'area_ur' 'complicaro' 'dia_cert' 'mes_cert'
# 'anio_cert' 'peso' 'lengua' 'cond_act' 'par_agre' 'ent_ocules'
# 'mun_ocules' 'loc_ocules' 'loc_resid' 'loc_ocurr' 'razon_m'] 
# Total:  59 